### PR TITLE
Fix serial number and firmware version on StreamDeck XL models in hs.streamdeck

### PR DESCRIPTION
**Pull request by cmsj**

### CODE REVIEW

File: extensions/streamdeck/HSStreamDeckDeviceXL.m [hunk 1]
```diff
@@ -58,17 +58,29 @@ - (NSString *)cacheSerialNumber {
...
-    return [[NSString alloc] initWithData:[self deviceRead:32 reportID:0x05]
-                                 encoding:NSUTF8StringEncoding];
+    NSUInteger fwOffset = 1;
+    NSData *firmwareVersionDataRaw = [self deviceRead:32 reportID:0x05];
+    NSRange offset = NSMakeRange(fwOffset, firmwareVersionDataRaw.length - fwOffset);
+    NSData *firmwareVersionData = [firmwareVersionDataRaw subdataWithRange:offset];
+
+    NSString *firmwareVersion =  [[NSString alloc] initWithData:firmwareVersionData
+                                                       encoding:NSUTF8StringEncoding];
+    return firmwareVersion;
```
Comment (jcutler) [inline extensions/streamdeck/HSStreamDeckDeviceXL.m:83]: Confirmed this change also works for Mk2

**Comment by jcutler** [comment]
Looks like this is the culprit for serial: https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/streamdeck/HSStreamDeckDevice.m#L80

At least on Mk2, serial is offset 2 from the beginning of the report, so we're going past it here. I think changing this to `report + 2` and `fwOffset` to 4, or just report and `fwOffset` to 6 and `serialOffset` to 2 would probably fix this.

Problems is I don't have the other devices to check the offsets for.

**Comment by cmsj** [comment]
Ok, I've done the refactoring I wanted to, so device subclasses now only have to set some variables about themselves, and implement `deviceWriteImage` if they don't follow the new "standard" way of doing it. I've copied your `initWithDevice` values into a new Mk2 class, but obviously I can't test it.

I've tested this on the Mini so far, but not the Original or XL.

**Comment by cmsj** [comment]
(Also I'd be happy to merge this without the Mk2 addition, so you can still propose the PR for that and get the git commit for it, if you'd like?)

**Comment by jcutler** [comment]
Tested this out. Setting `deviceReadOffset` to 2 results in a correct SN. But it has to be 6 to get the firmware version. Are you getting the correct output for both on the OriginalV2? I'd expect not.

**Comment by cmsj** [comment]
I don't have an OriginalV2 unfortunately. I guess we need to have different offsets for different commands. I can take a look at adding that tomorrow, but if you want to propose something, I can take a look in the morning. Thanks for helping nail this down!

**Comment by jcutler** [comment]
Yeah, the mini working for 5 makes sense: https://github.com/Julusian/node-elgato-stream-deck/blob/020a047dceb8816f5b884d1aef3de07482c5d8c3/packages/core/src/models/base.ts#L223-L232

The behaviour I'm seeing matches this: https://github.com/Julusian/node-elgato-stream-deck/blob/020a047dceb8816f5b884d1aef3de07482c5d8c3/packages/core/src/models/base-gen2.ts#L57-L66

Looks like Original and Mini are the first type, and OriginalV2, XL and Mk2 are the second type.

**Comment by jcutler** [comment]
I'm just going to embarrass myself a little bit:

1. It has been a while since I have contributed to another github project in a more complicated way than fork>commit change>PR, so I'm not sure how to keep context in this PR to show you my suggested change.
2. I almost never write ObjC, so I'm not sure if the `deviceReadOffset` param to `deviceRead` should be `CFIndex` or `NSUInteger`.

Sooo.... here's a patch for the changes that seemed to work for me:

```
diff --git a/extensions/streamdeck/HSStreamDeckDevice.h b/extensions/streamdeck/HSStreamDeckDevice.h
index 770ad8ab..bbf41f5f 100644
--- a/extensions/streamdeck/HSStreamDeckDevice.h
+++ b/extensions/streamdeck/HSStreamDeckDevice.h
@@ -49,7 +49,8 @@ typedef enum : NSUInteger {
 @property (nonatomic) int reportLength;
 @property (nonatomic) int reportHeaderLength;
 @property (nonatomic) int dataKeyOffset;
-@property (nonatomic) NSUInteger deviceReadOffset;
+@property (nonatomic) NSUInteger firmwareOffset;
+@property (nonatomic) NSUInteger serialNumberOffset;
 @property (nonatomic) NSData *resetCommand;
 @property (nonatomic) NSData *setBrightnessCommand;
 @property (nonatomic) NSUInteger serialNumberCommand;
@@ -67,7 +68,7 @@ typedef enum : NSUInteger {
 - (IOReturn)deviceWrite:(NSData *)report;
 - (void)deviceWriteImage:(NSData *)data button:(int)button;
 - (void)deviceV2WriteImage:(NSData *)data button:(int)button;
-- (NSData *)deviceRead:(int)resultLength reportID:(CFIndex)reportID;
+- (NSData *)deviceRead:(int)resultLength reportID:(CFIndex)reportID deviceReadOffset:(NSUInteger)deviceReadOffset;
 
 - (int)transformKeyIndex:(int)sourceKey;
 - (void)deviceDidSendInput:(NSArray*)newButtonStates;
diff --git a/extensions/streamdeck/HSStreamDeckDevice.m b/extensions/streamdeck/HSStreamDeckDevice.m
index d5600bb8..e4e8f78f 100644
--- a/extensions/streamdeck/HSStreamDeckDevice.m
+++ b/extensions/streamdeck/HSStreamDeckDevice.m
@@ -42,7 +42,8 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
         self.setBrightnessCommand = nil;
         self.serialNumberCommand = 0;
         self.firmwareVersionCommand = 0;
-        self.deviceReadOffset = 0;
+        self.firmwareOffset = 0;
+        self.serialNumberOffset = 0;
 
         serialNumberCache = nil;
         //NSLog(@"Added new Stream Deck device %p with IOKit device %p from manager %p", (__bridge void *)self, (void*)self.device, (__bridge void *)self.manager);
@@ -76,14 +77,14 @@ - (IOReturn)deviceWrite:(NSData *)report {
     return IOHIDDeviceSetReport(self.device, kIOHIDReportTypeFeature, rawBytes[0], rawBytes, report.length);
 }
 
-- (NSData *)deviceRead:(int)resultLength reportID:(CFIndex)reportID {
-    CFIndex reportLength = resultLength + self.deviceReadOffset;
+- (NSData *)deviceRead:(int)resultLength reportID:(CFIndex)reportID deviceReadOffset:(NSUInteger)deviceReadOffset {
+    CFIndex reportLength = resultLength + deviceReadOffset;
     uint8_t *report = malloc(reportLength);
 
     //NSLog(@"deviceRead: expecting resultLength %d, calculated report length %ld", resultLength, (long)reportLength);
 
     IOHIDDeviceGetReport(self.device, kIOHIDReportTypeFeature, reportID, report, &reportLength);
-    char *c_data = (char *)(report + self.deviceReadOffset);
+    char *c_data = (char *)(report + deviceReadOffset);
     NSData *dataRaw = [NSData dataWithBytes:c_data length:resultLength];
     free(report);
 
@@ -205,7 +206,7 @@ - (NSString *)cacheSerialNumber {
         return nil;
     }
 
-    NSData *serialNumberData = [self deviceRead:self.simpleReportLength reportID:self.serialNumberCommand];
+    NSData *serialNumberData = [self deviceRead:self.simpleReportLength reportID:self.serialNumberCommand deviceReadOffset:self.serialNumberOffset];
 
     NSString *serialNumber = [[NSString alloc] initWithData:serialNumberData
                                                    encoding:NSUTF8StringEncoding];
@@ -224,7 +225,7 @@ - (NSString*)firmwareVersion {
         [exception raise];
     }
 
-    NSString *firmwareVersion = [[NSString alloc] initWithData:[self deviceRead:self.simpleReportLength reportID:self.firmwareVersionCommand]
+    NSString *firmwareVersion = [[NSString alloc] initWithData:[self deviceRead:self.simpleReportLength reportID:self.firmwareVersionCommand deviceReadOffset:self.firmwareOffset]
                                                       encoding:NSUTF8StringEncoding];
     return firmwareVersion;
 }
diff --git a/extensions/streamdeck/HSStreamDeckDeviceMini.m b/extensions/streamdeck/HSStreamDeckDeviceMini.m
index 068d8f4e..615950bc 100644
--- a/extensions/streamdeck/HSStreamDeckDeviceMini.m
+++ b/extensions/streamdeck/HSStreamDeckDeviceMini.m
@@ -37,7 +37,8 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
         self.serialNumberCommand = 0x03;
         self.firmwareVersionCommand = 0x4;
 
-        self.deviceReadOffset = 5;
+        self.serialNumberOffset = 5;
+        self.firmwareOffset = 5;
     }
     return self;
 }
diff --git a/extensions/streamdeck/HSStreamDeckDeviceMk2.m b/extensions/streamdeck/HSStreamDeckDeviceMk2.m
index 38318f29..9c957749 100644
--- a/extensions/streamdeck/HSStreamDeckDeviceMk2.m
+++ b/extensions/streamdeck/HSStreamDeckDeviceMk2.m
@@ -35,6 +35,9 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
 
         self.serialNumberCommand = 0x06;
         self.firmwareVersionCommand = 0x05;
+
+        self.serialNumberOffset = 2;
+        self.firmwareOffset = 6;
     }
     return self;
 }
diff --git a/extensions/streamdeck/HSStreamDeckDeviceOriginal.m b/extensions/streamdeck/HSStreamDeckDeviceOriginal.m
index 78c3e8b6..0042c0e1 100644
--- a/extensions/streamdeck/HSStreamDeckDeviceOriginal.m
+++ b/extensions/streamdeck/HSStreamDeckDeviceOriginal.m
@@ -36,7 +36,8 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
         self.serialNumberCommand = 0x3;
         self.firmwareVersionCommand = 0x4;
 
-        self.deviceReadOffset = 5;
+        self.serialNumberOffset = 5;
+        self.firmwareOffset = 5;
     }
     return self;
 }
diff --git a/extensions/streamdeck/HSStreamDeckDeviceOriginalV2.m b/extensions/streamdeck/HSStreamDeckDeviceOriginalV2.m
index c1807c9d..b8d94305 100644
--- a/extensions/streamdeck/HSStreamDeckDeviceOriginalV2.m
+++ b/extensions/streamdeck/HSStreamDeckDeviceOriginalV2.m
@@ -35,6 +35,9 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
 
         self.serialNumberCommand = 0x06;
         self.firmwareVersionCommand = 0x05;
+
+        self.serialNumberOffset = 2;
+        self.firmwareOffset = 6;
     }
     return self;
 }
diff --git a/extensions/streamdeck/HSStreamDeckDeviceXL.m b/extensions/streamdeck/HSStreamDeckDeviceXL.m
index f7430ea3..26523179 100644
--- a/extensions/streamdeck/HSStreamDeckDeviceXL.m
+++ b/extensions/streamdeck/HSStreamDeckDeviceXL.m
@@ -35,6 +35,9 @@ - (id)initWithDevice:(IOHIDDeviceRef)device manager:(id)manager {
 
         self.serialNumberCommand = 0x06;
         self.firmwareVersionCommand = 0x05;
+
+        self.serialNumberOffset = 2;
+        self.firmwareOffset = 6;
     }
     return self;
 }
```

**Comment by cmsj** [comment]
thanks, I'll take a look

**Comment by jcutler** [comment]
Let me know if there's anything I can help with!

**Comment by jcutler** [comment]
I've got a Mk.2 I'd love to mess around with. Is this a PR you can find time for, @cmsj?

**Comment by cmsj** [comment]
I'm hoping to get it into shape at least for the Original, Mini and XL I have, this weekend.

**Comment by latenitefilms** [comment]
@cmsj - Ping me if there's any builds you need me to test later today.

**Comment by cmsj** [comment]
@jcutler thanks for the patch, that was very helpful!
I've now tested this on my Original, Mini and XL, so @jcutler if you could test the Mk2 that would be awesome.
@latenitefilms do you have an OriginalV2? I'm not sure that actually any of us do.

**Comment by latenitefilms** [comment]
I have the model **20GAA9901** - not sure if this is the original or the V2 - I think the original?

**Comment by cmsj** [comment]
yeah that's the Original, so that's the same one I have and have tested. Oh well, if anyone has an OriginalV2 and we're breaking it, I'm sure they'll let us know :D

**Comment by latenitefilms** [comment]
Is **10GAA9901** the V2?

**Comment by cmsj** [comment]
Hmm, I'm not sure. if it is, it would have a USB product ID of: 0x006d (109)

**Comment by latenitefilms** [comment]
When the Stream Deck Pedal eventually becomes available in Australia, I'll see if I can grab a V2 as well.

**Comment by cmsj** [comment]
I'm going to merge this so everyone gets a dev build to test with, but I won't cut a release until at least someone confirms it works on V2 (plus there are a few more issues I want to pick through before releasing 0.9.94)

**Comment by cmsj** [comment]
dev build will appear here: https://github.com/Hammerspoon/hammerspoon/actions/runs/1878886018

**Comment by jcutler** [comment]
LGTM on Mk2!

**Comment by cmsj** [comment]
thanks!

**Comment by jcutler** [comment]
IMG_20220228_005259
Thank you!

**Pull request merged**